File: training/chesskers/repetition.py
"""Position repetition tracking for draw-by-threefold (mirrors game-engine)."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .rules import Board


def _piece_key(piece) -> str:
    en_passant = "1" if piece.type == "pawn" and piece.en_passant else "0"
    return (
        f"{piece.x},{piece.y}:{piece.type}:{piece.team}:"
        f"{'1' if piece.has_moved else '0'}:{en_passant}"
    )


def position_key(board: Board) -> str:
    pieces = sorted(_piece_key(p) for p in board.pieces)
    hop = ""
    if board.checkers_hop_position is not None:
        hop = f"{board.checkers_hop_position[0]},{board.checkers_hop_position[1]}"
    return json.dumps(
        {"pieces": pieces, "totalTurns": board.total_turns, "hop": hop},
        separators=(",", ":"),
    )


def init_position_tracking(board: Board) -> Board:
    board.position_counts = {}
    board.is_draw = False
    board.position_counts[position_key(board)] = 1
    return board


def record_position(board: Board) -> None:
    if board.winning_team is not None or board.is_draw:
        return
    if not board.position_counts:
        init_position_tracking(board)
        return
    key = position_key(board)
    count = board.position_counts.get(key, 0) + 1
    board.position_counts[key] = count
    if count >= 3:
        board.is_draw = True


def is_terminal_board(board: Board) -> bool:
    return board.winning_team is not None or board.is_draw

File: engine/src/repetition.rs
use std::collections::HashMap;

use crate::board::Board;
use crate::state::PieceType;

fn piece_key(p: &crate::board::Piece) -> String {
    let en_passant = if p.piece_type == PieceType::Pawn && p.en_passant {
        "1"
    } else {
        "0"
    };
    format!(
        "{},{}:{:?}:{:?}:{}:{}",
        p.coord.x,
        p.coord.y,
        p.piece_type,
        p.team,
        if p.has_moved { "1" } else { "0" },
        en_passant
    )
}

pub fn position_key(board: &Board) -> String {
    let mut pieces: Vec<String> = board.pieces.iter().map(piece_key).collect();
    pieces.sort();
    let hop = board
        .checkers_hop_position
        .map(|c| format!("{},{}", c.x, c.y))
        .unwrap_or_default();
    serde_json::json!({
        "pieces": pieces,
        "totalTurns": board.total_turns,
        "hop": hop,
    })
    .to_string()
}

pub fn init_position_tracking(board: &mut Board) {
    board.position_counts = HashMap::new();
    board.is_draw = false;
    let key = position_key(board);
    board.position_counts.insert(key, 1);
}

pub fn record_position(board: &mut Board) {
    if board.winning_team.is_some() || board.is_draw {
        return;
    }
    if board.position_counts.is_empty() {
        init_position_tracking(board);
        return;
    }
    let key = position_key(board);
    let count = board.position_counts.get(&key).copied().unwrap_or(0) + 1;
    board.position_counts.insert(key, count);
    if count >= 3 {
        board.is_draw = true;
    }
}

pub fn is_terminal_board(board: &Board) -> bool {
    board.winning_team.is_some() || board.is_draw
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::board::Piece;
    use crate::state::{Coord, PieceType, Team};

    fn make_board(pieces: Vec<Piece>, total_turns: u32) -> Board {
        Board {
            pieces,
            total_turns,
            winning_team: None,
            checkers_hop_position: None,
            is_draw: false,
            position_counts: HashMap::new(),
        }
    }

    #[test]
    fn declares_draw_on_third_identical_position() {
        let mut board = make_board(
            vec![Piece {
                coord: Coord { x: 4, y: 0 },
                piece_type: PieceType::King,
                team: Team::White,
                has_moved: false,
                en_passant: false,
                possible_moves: Vec::new(),
            }],
            1,
        );
        init_position_tracking(&mut board);
        let key = position_key(&board);

        record_position(&mut board);
        assert!(!board.is_draw);
        assert_eq!(board.position_counts.get(&key), Some(&2));

        record_position(&mut board);
        assert!(board.is_draw);
        assert_eq!(board.position_counts.get(&key), Some(&3));
        assert!(is_terminal_board(&board));
    }
}
